### ISSUE TITLE
gt gff3 -sort: Specify sequence-region sort order

### ISSUE THREAD

**Issue by sjackman** [genometools/genometools#581]
My sequence regions have numerical identifiers. Is it possible to sort these numerically so that they don't sort like so:
`1 10 11 12 16 17 18 2 21 3 4 5 6 7 8 9`

They're in the correct order in the FASTA file. Perhaps the order could be specified based on that?

**Comment by gordon** [comment]
It's certainly possible, but I'm not sure that it would be a good idea. Right now the identifiers are sorted lexicographically with `strcmp(3)`. Sorting numerical identifiers numerically would lead to some corner cases which have to be considered:

- What happens if pure numerical (e.g, `1`) and non-numerical (e.g., `ctg123`) identifiers are mixed?
- What about identifiers which end with a numerical? (e.g, `seq1 seq2 seq10`)

It would also mean that we break the sortedness-property for GFF3 files which have been sorted before such a change. Files sorted before such a change would not be sorted anymore and vice versa.

I understand that it would be nice in some cases, but I doubt it would be worth the effort given the reasons above.

@satta: What's your take on this?

**Comment by satta** [comment]
While I agree in general with what @gordon said, I do think one should distinguish different use cases for sorting: 
 - lexicographical seqid sorting with topologically sorted features for each CC (to obtain a sorted file by GenomeTools definition, which allows efficient processing e.g. by `gt merge`and friends), and

- sorting for downstream use outside of GenomeTools. There are other examples in which GenomeTools' definition of sortedness deviates from others' definitions. For instance, tabix (http://samtools.sourceforge.net/tabix.shtml) does not like the fact in GenomeTools' sorting feature start positions _per line_ are not guaranteed to be strictly increasing (see #355). In this case, this was solved by adding an additional  `gt gff3` option `-sortlines` to trigger a different sorting strategy. One could do the same for numerical sorting here (for purely numerical seqids only, which would sort before all alphanumerical seqids).

Please note that using the order of sequences in an external FASTA file also leaves room for ambiguity -- what if a seqid does not appear in the FASTA file?

**Comment by gordon** [comment]
Good point @satta! Adding a numerical sorting option for outside use-cases seems like a good approach. @sjackman: What do you think of this suggestion?

Basing the order on the order given in the FASTA files seems to be to complicated in my opinion and leads to the corner case @satta mentioned.

**Comment by sjackman** [comment]
A numerical sorting option works for me. Thanks!

**Comment by satta** [comment]
Okay, I will look at this and implement this in an alternative output stream. Unfortunately just adapting the GtSortStream to use different GtCompare functions besides `gt_genome_node_compare()` is not really practical, as it is assumed in other places in the code that sorted streams are always sorted using  `gt_genome_node_compare()`. So I guess I will have to do this only for output.

**Comment by satta** [comment]
Done! Use `-sortnum` for numerical sorting.

**Comment by sjackman** [comment]
Works like a charm! Thanks! One small item. If both `-sortnum` and `-sortlines` require `-sort`, could these two options imply `-sort` without requiring it be explicitly included on the command line?

```
❯❯❯ gt gff3 -sortnum foo.gff
gt gff3: error: option "-sortnum" requires option "-sort"
```

**Comment by satta** [comment]
Done.

**Comment by standage** [comment]
Just to be clear, though: it'll be a few minutes until the automated tests
finish running so we can verify nothing has been broken and merge the patch.

**Comment by Garonenur** [comment]
merged, go ahead and try it.

**Comment by sjackman** [comment]
Works great. Thanks for your willingness and speed!

**Comment by sjackman** [comment]
Any plans for a stable release that would include this feature?

**Comment by satta** [comment]
I wouldn't mind a bugfix release... but if there's no absolute urgency I would wait until have implemented a few other things which could then go into the release as well (some additions to the Lua bindings).

**Comment by sjackman** [comment]
No huge urgency. I'm just more comfortable when my pipelines depend only on release features. It makes it more difficult to run the pipeline on other computers when one particular module needs to be pulled from GitHub.

**Comment by sjackman** [comment]
There appears to be a bug in `gt gff3 -sortnum` that can cause duplicate `##sequence-region` that does not seem to affect `-sort`. I ran the following command with `-sort` and `-sortnum` and here's the diff of the two:

```sh
gt gff3 -sort a.gff b.gff
gt gff3 -sortnum a.gff b.gff
```

```diff
--- a/foo.gff
+++ b/foo.gff
@@ -1,18 +1,41 @@
 ##gff-version   3
 ##sequence-region   1 1 1238593
+##sequence-region   1 279640 1229424
+##sequence-region   2 1 676873
+##sequence-region   2 23289 655908
+##sequence-region   3 1 369480
+##sequence-region   3 121114 342520
+##sequence-region   4 1 319932
+##sequence-region   4 9445 304205
+##sequence-region   5 1 308816
+##sequence-region   5 4 100091
+##sequence-region   6 1 591782
+##sequence-region   6 61475 457425
+##sequence-region   7 1 474387
+##sequence-region   7 585 446210
+##sequence-region   8 1 284015
+##sequence-region   8 58761 273946
+##sequence-region   9 1 183372
+##sequence-region   9 29678 176602
 ##sequence-region   10 1 150036
+##sequence-region   10 38061 127792
 ##sequence-region   11 1 237718
+##sequence-region   11 32388 144279
 ##sequence-region   12 1 131948
+##sequence-region   12 95482 96669
 ##sequence-region   13 1 131654
 ##sequence-region   14 1 116408
 ##sequence-region   15 1 103233
 ##sequence-region   16 1 107813
+##sequence-region   16 5899 85250
 ##sequence-region   17 1 150032
+##sequence-region   17 40228 133406
 ##sequence-region   18 1 87812
+##sequence-region   18 6247 8015
 ##sequence-region   19 1 62488
-##sequence-region   2 1 676873
 ##sequence-region   20 1 62468
 ##sequence-region   21 1 33754
+##sequence-region   21 6068 6349
 ##sequence-region   22 1 29686
 ##sequence-region   23 1 48977
 ##sequence-region   24 1 14473
@@ -21,7 +44,6 @@
 ##sequence-region   27 1 13983
 ##sequence-region   28 1 6571
 ##sequence-region   29 1 5367
-##sequence-region   3 1 369480
 ##sequence-region   30 1 5296
 ##sequence-region   31 1 4893
 ##sequence-region   32 1 4762
@@ -31,12 +53,6 @@
 ##sequence-region   36 1 2546
 ##sequence-region   37 1 2505
 ##sequence-region   38 1 2229
-##sequence-region   4 1 319932
-##sequence-region   5 1 308816
-##sequence-region   6 1 591782
-##sequence-region   7 1 474387
-##sequence-region   8 1 284015
-##sequence-region   9 1 183372
…
```

**Comment by satta** [comment]
Hmmm... interesting. Thanks -- did  your input file already contain the sequence regions or were they added automatically (e.g. with `-tidy`)? It would be very helpful if you could share your input file.

**Comment by sjackman** [comment]
Here's a small test case.

```
~ ❯❯❯ cat a.gff
##gff-version   3
##sequence-region   1 1 1238593
##sequence-region   10 1 150036
##sequence-region   2 1 676873
##sequence-region   20 1 62468
~ ❯❯❯ gt gff3 -sort a.gff a.gff
##gff-version   3
##sequence-region   1 1 1238593
##sequence-region   10 1 150036
##sequence-region   2 1 676873
##sequence-region   20 1 62468
~ ❯❯❯ gt gff3 -sortnum a.gff a.gff
##gff-version   3
##sequence-region   1 1 1238593
##sequence-region   1 1 1238593
##sequence-region   2 1 676873
##sequence-region   2 1 676873
##sequence-region   10 1 150036
##sequence-region   10 1 150036
##sequence-region   20 1 62468
##sequence-region   20 1 62468
```

**Comment by satta** [comment]
Thanks. Please find a fix in #585.

**Comment by sjackman** [comment]
Thanks, Sascha.
